File: documentation-examples-scala/src/scala/dev/typr/foundations/docs/core/OperationQueries.scala
package dev.typr.foundations.docs.core
import dev.typr.foundationssc.*

@SuppressWarnings(Array("unused"))
object OperationQueries:
  case class User(id: Int, name: String)

  val userCodec: RowCodec[User] = RowCodec.builder[User]()
    .field(PgTypes.int4)(_.id)
    .field(PgTypes.text)(_.name)
    .build(User.apply)

  val fragment: Fragment = Fragment.of("SELECT id, name FROM users")

  //start
  // Multi-column: pass a RowCodec with a result mode
  val allUsers: Operation[List[User]] = fragment.query(userCodec.all())
  val maybeUser: Operation[Option[User]] = fragment.query(userCodec.maxOne())
  val oneUser: Operation[User] = fragment.query(userCodec.exactlyOne())

  // Single-column: shorthand methods skip the codec
  val allIds: Operation[List[Int]] =
    Fragment.of("SELECT id FROM users").queryAll(PgTypes.int4)
  val maybeName: Operation[Option[String]] =
    Fragment.of("SELECT name FROM users LIMIT 1").queryMaxOne(PgTypes.text)
  val count: Operation[Int] =
    Fragment.of("SELECT count(*) FROM users").queryExactlyOne(PgTypes.int4)
  //stop

File: documentation-examples-java/src/java/dev/typr/foundations/docs/core/OperationQueries.java
package dev.typr.foundations.docs.core;

import dev.typr.foundations.*;
import java.util.List;
import java.util.Optional;

@SuppressWarnings("unused")
public class OperationQueries {
  record User(int id, String name) {}

  static RowCodec<User> userCodec =
      RowCodec.<User>builder()
          .field(PgTypes.int4, User::id)
          .field(PgTypes.text, User::name)
          .build(User::new);

  Fragment fragment = Fragment.of("SELECT id, name FROM users");

  // start
  // Multi-column: pass a RowCodec with a result mode
  OperationRead<List<User>> allUsers = fragment.query(userCodec.all());
  OperationRead<Optional<User>> maybeUser = fragment.query(userCodec.maxOne());
  OperationRead<User> oneUser = fragment.query(userCodec.exactlyOne());

  // Single-column: shorthand methods skip the codec
  OperationRead<List<Integer>> allIds = Fragment.of("SELECT id FROM users").queryAll(PgTypes.int4);
  OperationRead<Optional<String>> maybeName =
      Fragment.of("SELECT name FROM users LIMIT 1").queryMaxOne(PgTypes.text);
  OperationRead<Integer> count =
      Fragment.of("SELECT count(*) FROM users").queryExactlyOne(PgTypes.int4);
  // stop
}
